Pass through of target features to llvm-bitcode-linker and handling them

kulst opened the pull request:
When using the llvm-bitcode-linker (`linker-flavor=llbc`) target-features are not passed through and are not handled by it.
The llvm-bitcode-linker is mainly used as a self contained linker to link llvm bitcode for the nvptx64 target. It uses `llvm-link`, `opt` and `llc` internally. To produce a `.ptx` file of a specific ptx-version it is necessary to pass the version to llc with the `--mattr` option. Without explicitly setting it, the emitted `.ptx`-version is the minimum supported version of the `--target-cpu`. 

I would like to be able to explicitly set the ptx version as [some llvm problems only occur in earlier `.ptx`-versions](https://github.com/llvm/llvm-project/issues/112998).

Therefore this pull request adds support for passing target features to llvm-bitcode-linker and handling them. 
I was not quite sure if adding these features to `rustc_target/src/target_features.rs` is necessary or not. If so I will gladly add these.

    r? @kjetilkjeka

kulst committed:
Pass through of target features to llvm-bitcode-linker and handling them

The .ptx version produced by llc can be specified by passing it with --mattr. Currently it is not possible to specify the .ptx version with -Ctarget-feature because these are not passed through to llvm-bitcode-linker and handled by it. This commit adds both.
--target-feature and -mattr are passed with equals to mitigate issues when the value starts with a - (minus).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13168289218/job/36753488423) [(plain)](https://github.com/rust-lang/rust/commit/f4e1ac42a70653fe0fbeb62fb25f1c9395c9e63b/checks/36753488423/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 28.2s done
#21 DONE 41.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
```

</details>

workingjubilee commented:
r? @kjetilkjeka 

but I don't think it will work :thinking:

kjetilkjeka commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2518,6 +2518,12 @@ fn add_order_independent_options(
             "--target-cpu",
             &codegen_results.crate_info.target_cpu,
         ]);
+        if let Some(feat) = [sess.opts.cg.target_feature.as_str(), &sess.target.options.features]
```
I didn't have time to completely dig down into this topic. Since you seem to be actively working on it I will share what I have so far and if you don't complete my research before me rather look a bit deeper later.

I found it a bit hard to find clear documentation on how `--target-feature` is supposed to work. 

The part that seems fairly straight forward is that first of all `sess.target.option.cpu` being passed to llvm/llc will provide a list of standard features. Then  `sess.target.option.features` and `sess.opts.cg.target_features` is supposed to provide additions and subtractions to that using `+feat`/`-feat`.

I see that your implementation (and the implementation of bpf) replaces anything set in `sess.target.options.features` if `sess.opts.cg.target_features` is also set. Did you find a reference (RFC, code, etc) where it states that this is the intended behavior? It seems potentially surprising to me that the features that are considered default enabled by rustc is removed by adding user specified features while the ones that are considered default enabled by llvm requires explicit removal.

I did a quick codesearch and [this logic](https://github.com/rust-lang/rust/blob/59588250ad973ce69bd15879314c9769e65f36b3/compiler/rustc_codegen_llvm/src/llvm_util.rs#L344) is what I imagined would be the correct resolution for `target_feature`. I didn't get to follow the calls all the way down so I didn't really confirm that this is in fact the logic being used.

Assuming that this is the logic we actually want, perhaps it might make sense to share the implementation. I also didn't get to consider it to hard.

kulst replied:
I copied the implementation of bpf as it seemed to me it handles exactly the same purpose as the change in this pull request. As nvptx64 does not have any default target feature specified in its TargetOptions I was not too concerned that target features specified on the command line would overwrite these. It seemed to me as a good enough solution.

The logic you pointed out is definitely much better (maybe also for bpf). I will try to figure out how we can use this logic without duplicated code.

kulst commented:
Thanks for the feedback. 

I would prefer to add these features to `rustc_target/src/target_features.rs` in a different PR.

The behavior of llc that I could observe during my tests is the highest ptx version specified wins. The logic in llvm is generated by tablegen and looks like this:
```Cpp
if (Bits[NVPTX::PTX32] && PTXVersion < 32) PTXVersion = 32;
if (Bits[NVPTX::PTX40] && PTXVersion < 40) PTXVersion = 40;
// ...
if (Bits[NVPTX::PTX87] && PTXVersion < 87) PTXVersion = 87;
```
So `-mattr=+ptx85,+ptx78` produces ptx85.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13217881460/job/36899315557) [(plain)](https://github.com/rust-lang/rust/commit/8f353aeffc7b8724c5d0b0c2e093237ee1fe686e/checks/36899315557/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 29.3s done
#22 DONE 35.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui-fulldeps/codegen-backend/hotplug.rs#normal stdout ----

error in revision `normal`: test compilation failed although it shouldn't!
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "normal" "--check-cfg" "cfg(test,FALSE,normal,dep,bindep)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.normal" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021" "--target" "arm-unknown-linux-gnueabi" "--crate-type=lib" "--emit=link=-" "-Zcodegen-backend=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.normal/auxiliary/libthe_backend.so"
--- stderr -------------------------------
--- stderr -------------------------------
##[error]error: internal compiler error: compiler/rustc_middle/src/query/plumbing.rs:702:5: `tcx.global_backend_features(())` is not supported for this key;
                                hint: Queries can be either made to the local crate, or the external crate. This error means you tried to use it for one that's not supported.
                                If that's not the case, global_backend_features was likely never assigned to a provider function.

thread 'rustc' panicked at compiler/rustc_middle/src/query/plumbing.rs:702:5:
Box<dyn Any>
stack backtrace:
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic::BugAbort as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   3: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   4: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   5: rustc_middle::util::bug::bug_fmt
   6: rustc_middle::query::plumbing::default_query
   6: rustc_middle::query::plumbing::default_query
   7: <<rustc_middle::query::Providers as core::default::Default>::default::{closure#288} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, ())>>::call_once
      [... omitted 3 frames ...]
   8: <rustc_codegen_ssa::CrateInfo>::new
   9: <the_backend::TheBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  11: rustc_interface::passes::start_codegen
  12: <rustc_interface::queries::Linker>::codegen_and_build_linker
  12: <rustc_interface::queries::Linker>::codegen_and_build_linker
  13: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  14: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  15: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  16: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  18: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
  19: rustc_span::create_session_globals_then::<(), rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.86.0-nightly (b32cd5602 2025-02-08) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 --crate-type lib -Z codegen-backend=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.normal/auxiliary/libthe_backend.so
query stack during panic:
#0 [global_backend_features] computing the backend features for CLI flags
end of query stack
error: aborting due to 1 previous error
error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui-fulldeps/codegen-backend/hotplug.rs#bindep stdout ----

error in revision `bindep`: test compilation failed although it shouldn't!
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "bindep" "--check-cfg" "cfg(test,FALSE,normal,dep,bindep)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.bindep" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021" "--target" "arm-unknown-linux-gnueabi" "--crate-type=lib" "--emit=link,dep-info=-" "-Zbinary-dep-depinfo" "-Zcodegen-backend=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.bindep/auxiliary/libthe_backend.so"
--- stdout -------------------------------
/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.bindep/libhotplug.rlib: /checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.bindep/auxiliary/libthe_backend.so
/checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs:
/checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs:
/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.bindep/auxiliary/libthe_backend.so:
--- stderr -------------------------------
--- stderr -------------------------------
##[error]error: internal compiler error: compiler/rustc_middle/src/query/plumbing.rs:702:5: `tcx.global_backend_features(())` is not supported for this key;
                                hint: Queries can be either made to the local crate, or the external crate. This error means you tried to use it for one that's not supported.
                                If that's not the case, global_backend_features was likely never assigned to a provider function.

thread 'rustc' panicked at compiler/rustc_middle/src/query/plumbing.rs:702:5:
Box<dyn Any>
stack backtrace:
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic::BugAbort as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   3: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   4: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   5: rustc_middle::util::bug::bug_fmt
   6: rustc_middle::query::plumbing::default_query
   6: rustc_middle::query::plumbing::default_query
   7: <<rustc_middle::query::Providers as core::default::Default>::default::{closure#288} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, ())>>::call_once
      [... omitted 3 frames ...]
   8: <rustc_codegen_ssa::CrateInfo>::new
   9: <the_backend::TheBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  11: rustc_interface::passes::start_codegen
  12: <rustc_interface::queries::Linker>::codegen_and_build_linker
  12: <rustc_interface::queries::Linker>::codegen_and_build_linker
  13: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  14: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  15: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  16: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  18: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
  19: rustc_span::create_session_globals_then::<(), rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.86.0-nightly (b32cd5602 2025-02-08) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 --crate-type lib -Z binary-dep-depinfo -Z codegen-backend=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.bindep/auxiliary/libthe_backend.so
query stack during panic:
#0 [global_backend_features] computing the backend features for CLI flags
end of query stack
error: aborting due to 1 previous error
error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui-fulldeps/codegen-backend/hotplug.rs#dep stdout ----

error in revision `dep`: test compilation failed although it shouldn't!
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "dep" "--check-cfg" "cfg(test,FALSE,normal,dep,bindep)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.dep" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021" "--target" "arm-unknown-linux-gnueabi" "--crate-type=lib" "--emit=link,dep-info=-" "-Zcodegen-backend=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.dep/auxiliary/libthe_backend.so"
/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.dep/libhotplug.rlib: /checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs

/checkout/tests/ui-fulldeps/codegen-backend/hotplug.rs:
------------------------------------------
------------------------------------------
--- stderr -------------------------------
##[error]error: internal compiler error: compiler/rustc_middle/src/query/plumbing.rs:702:5: `tcx.global_backend_features(())` is not supported for this key;
                                hint: Queries can be either made to the local crate, or the external crate. This error means you tried to use it for one that's not supported.
                                If that's not the case, global_backend_features was likely never assigned to a provider function.

thread 'rustc' panicked at compiler/rustc_middle/src/query/plumbing.rs:702:5:
Box<dyn Any>
stack backtrace:
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic::BugAbort as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   3: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   4: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   5: rustc_middle::util::bug::bug_fmt
   6: rustc_middle::query::plumbing::default_query
   6: rustc_middle::query::plumbing::default_query
   7: <<rustc_middle::query::Providers as core::default::Default>::default::{closure#288} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, ())>>::call_once
      [... omitted 3 frames ...]
   8: <rustc_codegen_ssa::CrateInfo>::new
   9: <the_backend::TheBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  11: rustc_interface::passes::start_codegen
  12: <rustc_interface::queries::Linker>::codegen_and_build_linker
  12: <rustc_interface::queries::Linker>::codegen_and_build_linker
  13: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  14: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  15: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  16: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  18: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
  19: rustc_span::create_session_globals_then::<(), rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.86.0-nightly (b32cd5602 2025-02-08) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 --crate-type lib -Z codegen-backend=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/codegen-backend/hotplug.dep/auxiliary/libthe_backend.so
query stack during panic:
#0 [global_backend_features] computing the backend features for CLI flags
end of query stack
error: aborting due to 1 previous error
```

</details>

kulst commented:
I changed the implementation so that the existing `global_backend_features` query is used. This query calls [`global_llvm_features`](https://github.com/rust-lang/rust/blob/8ad2c9724d983cfb116baab0bb800edd17f31644/compiler/rustc_codegen_llvm/src/llvm_util.rs#L600) which to my knowledge is the current logic used to collect the target features that get passed to llvm.
To make this possible it is neccessary to query before link stage, as the `TyCtxt` is no longer present there. Therefore an additional field is added for the `CrateInfo` type to persist the target features until link stage.

Other options would have been:

1. Using `sess.target_features`: These are collected by `target_features_cfg` which does only incorporate features known to `rustc`. This would contradict with the current behavior of rustc to pass also `target_features` to llvm unknown to `rustc`.
2. Duplicating the logic that is used by either of the two functions into `rustc_codegen_ssa`: As we already have two functions in `rustc_codgen_llvm` [that use a similar logic](https://github.com/rust-lang/rust/issues/134792) I really did not want to do this.

kulst committed:
Persist target features used for codegen beyond tcx

Bitcode linkers like llvm-bitcode-linker or bpf linker hand over the target features to llvm during link stage. During link stage the `TyCtxt` is already gone so it is not possible to create a query for the global backend features any longer. The features preserved in `Session.target_features` only incorporate target features known to rustc. This would contradict with the behaviour during codegen stage which also passes target features to llvm which are unknown to rustc.
This commit adds target features as a field to the `CrateInfo` struct and queries the target features in its new function. This way the target features are preserved beyond tcx and available at link stage.
To make sure the `global_backend_features` query is always registered even if the CodegenBackend does not register it, this registration is added to the `provide`function of the `rustc_codegen_ssa` crate.

kulst commented:
I slightly changed the implementation. `rustc_codegen_ssa` acts now as a provider for the default `global_backend_features` query. This ensures the query is always registered even if a `CodegenBackend` does not register it by its `provide` method.
I am now quite happy with the implementation and am looking forward getting additional feedback.

kjetilkjeka commented:
It looks like you found the right place to hook onto the target_feature. 

It seems to me like `CrateInfo` is an old name and since it was merged with `LinkerInfo` it has also contained fields like `target_cpu`. Based on that I agree that it's also appropriate to include the `target_features` :+1: 

I'm not an approved reviewer for rustc, but it looks great from my perspective. If I could I would have r+'ed.

For the follow up, where you add the ptx features. Also add a couple of tests where we check that using features like `ptx85` and/or `ptx75` gives the correct behavior in ptx. I see that there's currently no tests for `global_llvm_features` and a comment that express a bit uncertainty about the exact precedence.

Thanks a lot for contributing to the `llvm-bitcode-linker` and also for help improving the experience of using the `nvptx` in Rust. I greatly look forward to unlock everything that comes with the debug symbols. :heart:

kulst commented:
Thanks again for the review and the positive feedback! 
I will take your comments into consideration for the follow up.
I wasn't sure if you are allowed to officially review rustc pull requests as an nvptx target maintainer. But since I was interested in your feedback anyway, I decided to give it a try.

bjorn3 commented on src/tools/llvm-bitcode-linker/src/linker.rs:
```diff
@@ -134,6 +141,10 @@ impl Session {
         }
 
+        if let Some(mattr) = &self.feature {
+            lcc_command.arg(&format!("-mattr={}", mattr));
```
Is there a reason this uses `-mattr=...` while the above code does `--mcpu ...`?

kulst replied:
No particular reason, is now adjusted

bjorn3 commented on compiler/rustc_codegen_ssa/src/traits/backend.rs:
```diff
@@ -136,3 +136,10 @@ pub trait ExtraBackendMethods:
+        global_backend_features: |_tcx: TyCtxt<'_>, ()| vec![],
+        ..*providers
+    };
+}
```
This should not be in the traits module. The only code in the traits module should be definitions of traits. Maybe put it directly in the `provide` function of lib.rs?

kulst replied:
done

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 831d9f39e9c330934b7c4f35010477ba4a738d96 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 831d9f39e9c330934b7c4f35010477ba4a738d96 -->
<!-- homu: {"type":"Approved","sha":"831d9f39e9c330934b7c4f35010477ba4a738d96","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
